src/main/java/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -1197,18 +1214,97 @@ public SSLEngineResult unwrap(ByteBuffer src, ByteBuffer[] dsts, int offset, int
+                //  - Up to 2^16 - 1 bytes of content.
+                //
+                // Fetch and validate the ContentType as a sanity check that
+                // we're at the start of a record.
+                byte fragment_type = src.get();
+                validate_content_type(fragment_type);
+
+                // Ignore the legacy protocol version; we leave it to NSS to
+                // validate this for us.
+                src.get(); src.get();
+
+                // Get and validate the fragment's length against remaining
+                // data in src.
+                int fragment_length = src.getShort();
+                if (src.remaining() < fragment_length) {
+                    // Since we have too little data remaining in our
+                    // fragment, reset to the start of the fragment and
+                    // return.
+                    src.position(fragment_start);
+                    return new SSLEngineResult(SSLEngineResult.Status.BUFFER_UNDERFLOW, handshake_state, wire_data, app_data);
+                }
+
+                // We know we have a complete record fragment in src. This
+                // means we can compute the total bounds of this fragment:
+                int fragment_end = fragment_start + 5 + (int)fragment_length;
+                // and can read it into fragment_data. Note that we have to
+                // reset src's position so we get the fragment header as well.
+                fragment_data = new byte[fragment_end - fragment_start];
+                src.position(fragment_start);
+                src.get(fragment_data);
+
+                debug("JSSEngine.unwrap(): Found record fragment of size " + fragment_data.length + " bytes.");
+            }
+
+            if (fragment_data != null) {
+                // We have previously validated fragment data. Copy it into
+                // our write buffers. First compute the size of this write:
+                this_src_write = Math.min((int) Buffer.WriteCapacity(read_buf), fragment_data.length);
 
-                // When we have data from src, write it to read_buf.
                 if (this_src_write > 0) {
-                    byte[] wire_buffer = new byte[this_src_write];
-                    src.get(wire_buffer);
+                    // Create a temporary buffer containing the bytes we know
+                    // will be read into the buffer.
+                    byte[] wire_buffer = Arrays.copyOfRange(fragment_data, 0, this_src_write);
+
+                    // Update our fragment_data, either setting it to null or
+                    // removing the bytes we've read.
+                    if (this_src_write < fragment_data.length) {
+                        fragment_data = Arrays.copyOfRange(fragment_data, this_src_write, fragment_data.length);
+                    } else {
+                        fragment_data = null;
+                    }
 
+                    // Perform the underlying byte buffer call.
```

Comment:
We don't exit while `this_src-write > 0`, but if NSS gets stuck with a partial record fragment, I think perhaps we should reset `src`... Let me update that.